$\text{IMPORTS}$

In [4]:
import cvxpy as cp
import numpy as np
import matplotlib.pyplot as plt
import scipy.optimize as spo

$\textbf{QUESTION 1}$

$$
\text{PART A}\\~\\
x = [1, 2, 3], y = [3, 2, 1] \Rightarrow \\
x - y = -2x + 0y -2z
$$

<p align="center">
<img src="/home/leonardo_leads/Documents/SchoolDocs/ga_tech_masters/omscs_ml/fall2023/deterministic_optimization/lectures/week12Modules23and24/homework/images/1a.png" width="600" height="600" />
</p>

$$
\text{No, it's not true that } x \succ_\mathbb{R^{3}_{+}} y, \text{ because the line is outside the orthant}
$$

$$
\text{PART B}\\~\\
x = [1, 2, 3], y = [3, 2, 1] \Rightarrow \\
x - y = -2x + 0y -2z
$$

<p align="center">
<img src="/home/leonardo_leads/Documents/SchoolDocs/ga_tech_masters/omscs_ml/fall2023/deterministic_optimization/lectures/week12Modules23and24/homework/images/1b.png" width="600" height="600" />
</p>

$$
\text{Yes, it's true that } x \succ_\mathbb{L^{3}} y, \text{ because the line is inside the cone}
$$

$$
\text{PART C}\\
$$

In [22]:
A = [[-6, 7, 8], [7, -8, 9], [8, 9, -10]]
B = [[0, 1, 2], [1, 2, 3], [2, 3, 4]]
C = [[-10, 9, 8], [9, -8, 7], [8, 7, -12]]

A = np.array(A)
B = np.array(B)
C = np.array(C)

eigBA = np.linalg.eig(B - A).eigenvalues
eigAC = np.linalg.eig(A - C).eigenvalues

print(f"Eigenvalues of B - A:\n{eigBA}")
print(f"Eigenvalues of A - C:\n{eigAC}")

Eigenvalues of B - A:
[-2.58300524 14.         18.58300524]
Eigenvalues of A - C:
[-1.75877048  5.06417777  2.69459271]


$$
\text{No, it's false that } A \prec_\mathbb{S^{3}_{+}} B, \text{ because of the negative eigenvalue, } \approx -2.58 \Rightarrow\\ 
\text{This means we don't have a PSD matrix.}\\~\\

\text{No, it's false that } A \succ_\mathbb{S^{3}_{+}} C, \text{ because of the negative eigenvalue, } \approx -1.75 \Rightarrow \\
\text{This means we don't have a PSD matrix.}
$$

$\textbf{QUESTION 2}$

In [26]:
uni = np.random.uniform(0, 10, (20, 2))
fire_station = cp.Variable(2)
euclid_dist = cp.Variable()

min_dis = cp.Minimize(euclid_dist)

constraints = [cp.norm(uni[i] - fire_station) <= euclid_dist for i in range(10)]

fire2village = cp.Problem(min_dis, constraints)
fire2village.solve()

station_loc1 = fire_station.value[0]
station_loc2 = fire_station.value[1]

min_long_dist = fire2village.value

In [30]:
print(f"Fire Station Optimal Location:\n{station_loc1=}, {station_loc2=}")
print(f"\nMinimum Longest Distance, Fire Station to Village:\n{min_long_dist}")

Fire Station Optimal Location:
station_loc1=4.368366578875536, station_loc2=5.2083293463025155

Minimum Longest Distance, Fire Station to Village:
4.928239934911055
